### PR TITLE
Move from sax to saxes for XML parsing

### PR DESCRIPTION
**Pull request by lddubeau**
## Motivation

Sax does not appear to be actively developed anymore, which means that request for fixes or features go unanswered. The [PR filed](https://github.com/isaacs/sax-js/pull/231) to fix the issue reported in #2213 has gone unanswered for over 3 months.  See also [this issue](https://github.com/isaacs/sax-js/issues/238) asking whether the project is still alive.

I recently forked saxes from sax for my own XML parsing purposes, and fixed a lot of issues with how sax parses XML. 

## Benefits for JSDOM

* Better conformance with the XML specification. ([XML 1.0 fifth edition](https://www.w3.org/TR/2008/REC-xml-20081126/) and [XML Namespaces 1.0 third edition](http://www.w3.org/TR/2009/REC-xml-names-20091208/), to be specific.) Known limitation regarding conformance: saxes does not do well-formedness checks *inside* the DTD that may be included in a document. It is rare nowadays to rely on DTDs when doing XML processing in browser environments. Adding code to check that the DTD is well-formed would require *a lot of work* for very little benefit.

* Parsing behavior more closely aligned with those browsers that people normally use (like Chrome and Firefox.) The issue with empty comments and empty CDATA sections is one example: JSDOM+sax would not create nodes for `<!---->` and `<![CDATA[]]>` because sax did not report them. JSDOM+saxes produces nodes for both, which is consistent with other browsers.

## Breaking Changes

Prior to this PR, JSDOM accepted malformed XML. A prime example is the empty string. An empty string is not a well-formed XML document. Prior to this PR you could do `new DOMParser().parseFromString("", "text/xml")` and get an unpopulated document. After this PR, this is not possible anymore: you get an error. (This is consistent with how Chrome and Firefox behave.) So users may have been able to run JSDOM with malformed XML documents in the past. They won't be able to do this anymore.

## Other Considerations

I've noticed that prior to this PR, JSDOM was unable to properly handle `<script><![CDATA[ ... ]]></script>`. It so happens that a lot of the tests in web-platform-test suite that are in XML files have to make use of CDATA sections in `script` elements, so that `<` and `&` don't need escaping. Prior to this PR, upon encountering such `script` construct, JSDOM would build an `HTMLScriptElement` containing literally `<![CDATA[...`, which obviously is not valid JavaScript and this would cause the test to immediately fail. This made quite a few tests that are part of the WPT fail. This PR fixes the immediate problem of handling CDATA in `script`. There were a few tests that immediately became runnable after the parsing changes were made. This was reported by the test runner and I've turned on these tests (edited `to-run.yaml`). However, I've not reviewed the rest of the tests that were failing due to CDATA in `script`. There may be some tests in the lot that are failing for some easily fixable reason.

### CODE REVIEW

File: lib/jsdom/browser/htmltodom.js [hunk 1]
```diff
@@ -67,90 +67,89 @@ module.exports = class HTMLToDOM {
...
-          localName = prefix;
-          prefix = null;
-        }
+      for (const key of Object.keys(tag.attributes)) {
+        const attr = tag.attributes[key];
 
-        if (prefix === "") {
+        let { prefix, local } = attr;
+        const { uri, value } = attr;
+        if (prefix === "xmlns" && local === "") {
```
Comment (domenic) [inline lib/jsdom/browser/htmltodom.js]: Any interest in changing saxes to fix this bug?

Comment (lddubeau) [reply lib/jsdom/browser/htmltodom.js]: I don't know that I'd call it a "bug". Though I did call it a "peculiarity" in the comment. (I reworded an existing comment but "peculiarity" is my choice of words :smiley:  .) Looking at the commit history of sax, it seems to me it was a conscious choice to do it this way. There's some logic to it, and there are use-case scenarios where the current behavior is quite handy. Then again, as a "peculiarity", it violates the principle of least surprise, and it is not concordant with the XML Infoset. (Not that the Infoset drives how saxes should do things, but again: principle of least surprise.) So I guess the balance weighs in favor of removing that peculiar behavior.

Comment (lddubeau) [reply lib/jsdom/browser/htmltodom.js]: I've released a new version of saxes and amended this PR.

**Comment by Zirro** [comment]
Great to see the many failures due to CDATA-in-script fixed! Unfortunately, the previously quick fail in `/dom/nodes/ParentNode-querySelector-All-xht.xht` seems to have turned into a one minute timeout instead and its failure reason should be changed to reflect this.

**Comment by lddubeau** [comment]
@Zirro I've amended the last commit to include the change you pointed out.

Two questions come to my mind though:

1. Before marking `ParentNode-querySelector-All-xht.xht` as a timeout, I reran it and it *passed*, after something like 60 seconds. In every single test suite I'm responsible for, a timeout is a failure. Is there some special situation here that requires handling timeouts the way they are handled?

2. How can I detect cases like `ParentNode-querySelector-All-xht.xht`. Can I run the suite with a flag that would immediately point these out? Or do I have to look at each and every test in the test report and look for those that may happen to be timing out?

**Comment by Zirro** [comment]
Those are good questions. In this case the passing status actually means that the test failed as expected. We recently started running tests that are known to fail quickly in order to see if that's still the case after any given change. Considering the large amount of WPTs and their complexity, it's actually been pretty helpful in finding unexpected passing tests in various directories.

While we may be a bit limited by Mocha in the ways we can express "expected failures" during the run, we could certainly look into some ways of making it clearer that the tests have indeed still failed.

As for timeouts, we separate them from other failures by giving them the `timeout` status since they would take too long to run every time. The process could do with some refinement though - much of it can likely be automated later on.

> Or do I have to look at each and every test in the test report and look for those that may happen to be timing out?

It depends on the cause of the timeout. Some kinds are caught by `testharness.js` and result in an error message regarding a timeout at the end of the run, while others are skipped by the test framework after too much time has passed. These can usually be found by looking for their run time printed in red next to their names in the test log, though it may still take some work to separate them from tests which actually do pass but do so very slowly.

**Comment by domenic** [comment]
> In this case the passing status actually means that the test failed as expected.

We should change the "test name" in the harness output to something like "[EXPECTED FAIL] .../filename/...

**Pull request merged**